File: src/shell.rs
// shell — 비동기 커널 셸
//
// executor가 돌리는 async 태스크. 키보드 스트림을 await로 받아 한 줄을
// 조립하고, Enter에서 명령을 실행한다. 다른 태스크와 동시에 진행된다.

use crate::task::keyboard::ScancodeStream;
use crate::vga_buffer;
use crate::{allocator, interrupts, print, println};
use alloc::string::String;
use futures_util::stream::StreamExt;
use pc_keyboard::{layouts, DecodedKey, HandleControl, Keyboard, ScancodeSet1};

pub async fn run() {
    let mut scancodes = ScancodeStream::new();
    let mut keyboard = Keyboard::new(ScancodeSet1::new(), layouts::Us104Key, HandleControl::Ignore);
    let mut line = String::new();

    print_banner();
    print_prompt();

    while let Some(scancode) = scancodes.next().await {
        if let Ok(Some(event)) = keyboard.add_byte(scancode) {
            if let Some(DecodedKey::Unicode(c)) = keyboard.process_keyevent(event) {
                match c {
                    '\n' => {
                        println!();
                        execute(line.trim());
                        line.clear();
                        print_prompt();
                    }
                    '\u{8}' => {
                        if line.pop().is_some() {
                            vga_buffer::backspace();
                        }
                    }
                    c => {
                        line.push(c);
                        print!("{}", c); // 입력 에코
                    }
                }
            }
        }
    }
}

fn print_prompt() {
    print!("hobby_kernel> ");
}

fn print_banner() {
    println!();
    println!("  welcome to the hobby_kernel shell");
    println!("  type 'help' to see commands");
    println!();
}

fn execute(line: &str) {
    if line.is_empty() {
        return;
    }
    let mut parts = line.splitn(2, ' ');
    let cmd = parts.next().unwrap_or("");
    let args = parts.next().unwrap_or("");

    match cmd {
        "help" => {
            println!("commands:");
            println!("  help         show this help");
            println!("  echo <text>  print text back");
            println!("  about        about this OS");
            println!("  uptime       timer ticks since boot");
            println!("  mem          heap usage");
            println!("  clear        clear the screen");
            println!("  whoami       who is running this");
        }
        "echo" => println!("{}", args),
        "about" => {
            println!("hobby_kernel v0.1.0  --  a hobby OS written in Rust");
            println!("boot -> vga -> interrupts -> heap -> async tasks -> shell");
        }
        "uptime" => {
            // 타이머 인터럽트 발생 횟수. 실제 주파수는 설정에 따라 달라
            // 정확한 초 환산은 하지 않고 raw 틱을 그대로 보여준다.
            println!("timer ticks since boot: {}", interrupts::ticks());
        }
        "mem" => {
            let (used, free, total) = allocator::heap_stats();
            println!("heap: {} used / {} free / {} total (bytes)", used, free, total);
        }
        "clear" => vga_buffer::clear_screen(),
        "whoami" => println!("a proud weeb running their own OS :)"),
        _ => println!("unknown command: '{}'  (try 'help')", cmd),
    }
}

File: src/shell.c
// shell.c — 커널 셸
// UART 인터럽트가 shell_input(c)로 글자를 하나씩 넘긴다. 줄을 모아 Enter에서 실행.

#include "shell.h"
#include "uart.h"
#include "trap.h"  // clock_ticks()

static char line[128];
static int  len = 0;

// --- 작은 문자열 헬퍼 (libc 없음) ---
static int str_eq(const char *a, const char *b) {
    while (*a && *b) {
        if (*a != *b) return 0;
        a++; b++;
    }
    return *a == *b;
}
static int starts_with(const char *s, const char *p) {
    while (*p) {
        if (*s != *p) return 0;
        s++; p++;
    }
    return 1;
}

static void prompt(void) { uart_puts("hobby> "); }

static void execute(const char *cmd) {
    if (cmd[0] == '\0') {
        return;
    } else if (str_eq(cmd, "help")) {
        uart_puts("commands: help, about, uptime, clear, whoami, echo <text>\n");
    } else if (str_eq(cmd, "about")) {
        uart_puts("hobby-kernel (C / RISC-V) -- a from-scratch learning kernel\n");
    } else if (str_eq(cmd, "uptime")) {
        uart_puts("uptime: ");
        uart_dec(clock_ticks() / 10);  // 타이머 0.1초 간격 → /10 = 초
        uart_puts(" sec\n");
    } else if (str_eq(cmd, "clear")) {
        uart_puts("\033[2J\033[H");  // ANSI: 화면 지우고 커서 홈으로
    } else if (str_eq(cmd, "whoami")) {
        uart_puts("a kernel hacker building their own OS\n");
    } else if (starts_with(cmd, "echo ")) {
        uart_puts(cmd + 5);
        uart_putc('\n');
    } else {
        uart_puts("unknown command: ");
        uart_puts(cmd);
        uart_puts("  (try 'help')\n");
    }
}

void shell_init(void) {
    uart_puts("\ntype 'help' for commands.\n");
    prompt();
}

void shell_input(char c) {
    if (c == '\r' || c == '\n') {
        uart_putc('\n');
        line[len] = '\0';
        execute(line);
        len = 0;
        prompt();
    } else if (c == 0x7f || c == 0x08) {  // DEL / Backspace
        if (len > 0) {
            len--;
            uart_puts("\b \b");  // 화면에서도 한 글자 지움
        }
    } else if (len < (int)sizeof(line) - 1) {
        line[len++] = c;
        uart_putc(c);  // 입력 에코
    }
}
